### DISCUSSION TITLE
How do deployment requests work in production, when production has preexisting data?

### DISCUSSION THREAD

**Discussion by toddjudd** [planetscale/discussion#226] [Q&A] [answered]
I've only been working with planet scale and prisma for about a week now reading the docs and following some tutorials. Everything was going fine until I promoted a branch to production and then couldn't change the table schemas any more. I'm sure there's something I'm missing, or something I don't understand so assuming you have a main branch in production with data, what's the expected workflow for say adding a column? Here's what I've done:

### Main Schema
I had a db with a main branch in production with this schema and 5 or so records.
```
CREATE TABLE `PollQuestion` (
	`id` varchar(191) NOT NULL,
	`question` varchar(5000) NOT NULL,
	`endsAt` datetime(3),
	`options` json NOT NULL,
	`ownerToken` varchar(255) NOT NULL,
	`createdAt` datetime(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3),
	`updatedAt` datetime(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3),
	PRIMARY KEY (`id`),
  KEY `PollQuestion_ownerToken_idx` (`ownerToken`)
) ENGINE InnoDB,
  CHARSET utf8mb4,
  COLLATE utf8mb4_unicode_ci;
```

I created a branch using `pscale branch create main add-column-test` and connected to the branch using `pscale connect polls add-column-test`.
I then updated my `schema.prisma` file and used `npx prisma db push` to push to the `add-column-test` branch and created a deployment request using the pscale ui.

### Schema Change
This was the diff or schema changes shown in the ui.
```
 CREATE TABLE `PollQuestion` (
  `id` varchar(191) NOT NULL,
  `question` varchar(5000) NOT NULL,
  `endsAt` datetime(3),
  `options` json NOT NULL,
  `ownerToken` varchar(255) NOT NULL,
  `createdAt` datetime(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3),
  `updatedAt` datetime(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3),
+ `testToken` varchar(255) NOT NULL,
  PRIMARY KEY (`id`),
- KEY `PollQuestion_ownerToken_idx` (`ownerToken`)
+ KEY `PollQuestion_ownerToken_testToken_idx` (`ownerToken`, `testToken`)
 ) ENGINE InnoDB,
  CHARSET utf8mb4,
  COLLATE utf8mb4_unicode_ci;
```

### Main Schema After Deploy Request
in the pscale ui, everything looks correct the path database>branches>main>schema is correct.
```
CREATE TABLE `PollQuestion` (
	`id` varchar(191) NOT NULL,
	`question` varchar(5000) NOT NULL,
	`endsAt` datetime(3),
	`options` json NOT NULL,
	`ownerToken` varchar(255) NOT NULL,
	`createdAt` datetime(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3),
	`updatedAt` datetime(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3),
	`testToken` varchar(255) NOT NULL, 
	PRIMARY KEY (`id`),
	KEY `PollQuestion_ownerToken_testToken_idx` (`ownerToken`, `testToken`)
) ENGINE InnoDB,
  CHARSET utf8mb4,
  COLLATE utf8mb4_unicode_ci;
```

### Main Tables After Deploy Request
However looking at the table via the shell, prisma studio or my app i show 2 tables not 1
```
show tables;
4 rows in (4 ms)
Tables_in_polls
PollQuestion
_0cce185e_a13f_5225_b8f1_5185b1582561_20220703002403_vrepl
```

### Poll Question Definition
This is the table my app and prisma is quering from and I expect to have a new column here the the proper index
```
describe PollQuestion;

7 rows in (8 ms)
Field	Type	Null	Key	Default	Extra
id	varchar(191)	NO	PRI		
question	varchar(5000)	NO			
endsAt	datetime(3)	YES			
options	json	NO			
ownerToken	varchar(255)	NO	MUL		
createdAt	datetime(3)	NO		CURRENT_TIMESTAMP(3)	DEFAULT_GENERATED
updatedAt	datetime(3)	NO		CURRENT_TIMESTAMP(3)	DEFAULT_GENERATED
```

### _0cce185e_a13f_5225_b8f1_5185b1582561_20220703002403_vrepl
I do not understand why this table is created but it has no data. 
```
describe _0cce185e_a13f_5225_b8f1_5185b1582561_20220703002403_vrepl;

8 rows in (4 ms)
Field	Type	Null	Key	Default	Extra
id	varchar(191)	NO	PRI		
question	varchar(5000)	NO			
endsAt	datetime(3)	YES			
options	json	NO			
ownerToken	varchar(255)	NO	MUL		
createdAt	datetime(3)	NO		CURRENT_TIMESTAMP(3)	DEFAULT_GENERATED
updatedAt	datetime(3)	NO		CURRENT_TIMESTAMP(3)	DEFAULT_GENERATED
testToken	varchar(255)	NO			
```

What am I missing here and What is pscale doing here? Why is there a new table, and how should I be using?

**Comment by jgreet** [comment, marked as answer]
You're using the correct workflow for adding a column. Once a branch has been promoted to production status it is protected from direct DDL and all schema changes must be made using deploy requests.

The basics of that workflow are introduced here: 

https://docs.planetscale.com/concepts/planetscale-workflow
https://docs.planetscale.com/concepts/nonblocking-schema-changes

>**Poll Question Definition**
>
>This is the table my app and prisma is quering from and I expect to have a new column here the the proper index
>
> ```
> describe PollQuestion;
> 
> 7 rows in (8 ms)
> Field	Type	Null	Key	Default	Extra
> id	varchar(191)	NO	PRI		
> question	varchar(5000)	NO			
> endsAt	datetime(3)	YES			
> options	json	NO			
> ownerToken	varchar(255)	NO	MUL		
> createdAt	datetime(3)	NO		CURRENT_TIMESTAMP(3)	DEFAULT_GENERATED
> updatedAt	datetime(3)	NO		7 rows in (8 ms)
> ```
> 

That's definitely confusing. Once the deploy request has completed you should have the new column and updated index in your `main` branch. If I check the schema for your `main` branch it does appear to include those changes now. Did you run the `describe` before the deploy request finished?

>_0cce185e_a13f_5225_b8f1_5185b1582561_20220703002403_vrepl
>
>I do not understand why this table is created but it has no data.
>
>What am I missing here and What is pscale doing here? Why is there a new table, and how should I be using?

Those temporary tables are used during the online schema change process. They make it possible to transform tables with existing data without locking them or taking them offline. Often one of these tables is created with the new schema, existing rows and updates from the current table are streamed in to it, and when the new table is up to date it's swapped in for the old one.

A much better and longer description is available here:
https://docs.planetscale.com/learn/how-online-schema-change-tools-work

These temporary tables don't count against database storage limits and they'll be cleaned up automatically some time after the schema change has completed.

**Reply by toddjudd** [reply]
Thanks for the link and explanation regarding the temp tables, I honestly thought I was doing something wrong. I really the thoughtful response. I am, however, still stuck. 

You're correct the schema for the `main` branch is correct. It shows the `testToken` column and index as expected. Even now though, running `pscale deploy-request show polls 1` shows the status of the deploy request as `complete`. Still i'm seeing the temp tables. In fact there are now two that appear to be for the `PollQuestion` table, neither of which have data. Yet the `PollQuestion` table it's self remains unchanged both in description and it's data. 

Finally when trying to connect to the DB via prisma, using studio or via prisma in Next.js, I'm getting the following error code.
```
Message: Error in Prisma Client request: 


Invalid `prisma.pollQuestion.findMany()` invocation:


  The column `(not available)` does not exist in the current database.
  
Query:
{
  "modelName": "PollQuestion",
  "operation": "findMany",
  "args": {
    "take": 100,
    "skip": 0,
    "select": {
      "id": true,
      "question": true,
      "endsAt": true,
      "options": true,
      "ownerToken": true,
      "testToken": true,
      "createdAt": true,
      "updatedAt": true,
      "Vote": true
    }
  }
}
```

Is there another step after deployment that I'm not taking to finish the move of data via the temp tables to their correct format? Or is it possible this is just a bug?

**Reply by toddjudd** [reply]
I'm curious if this is related to this issue is related to this thread about [prisma migrate vs prisma db push](https://github.com/prisma/prisma/issues/10771#issuecomment-1065935613) though I've been reading in the [docs](https://docs.planetscale.com/tutorials/prisma-quickstart#push-prisma-schema-to-planetscale) that prisma migrate is no longer needed and that pscale should be covering what prisma migrate was responsible for.

**Reply by toddjudd** [reply]
This also may be because the column added is `not null` and there's no logical way to migrate the data from the original schema to the new one... If that's the case though, there's no apparent error or log stating that.

**Reply by jgreet** [reply]
>This also may be because the column added is not null and there's no logical way to migrate the data from the original schema to the new one... If that's the case though, there's no apparent error or log stating that.

After digging in our logs I think this is correct. The migration failed with this error:

` Field 'testToken' doesn't have a default value (errno 1364)`

But the deploy request was erroneously marked successful. We're fixing that bug now but if you retry the migration with a default value, or without the not null constraint, I think it will work.

Sorry for the confusion that caused.

The temporary tables are unrelated to this problem and will be garbage collected eventually.
